fuel-tx/src/transaction/fee.rs

```diff
@@ -142,57 +92,122 @@ impl TransactionFee {
         tx: &T,
     ) -> Option<Self>
     where
-        T: Chargeable + field::Inputs,
+        T: Chargeable,
     {
-        let metered_bytes = tx.metered_bytes_size() as Word;
-        let gas_used_by_metadata = tx.gas_used_by_metadata(gas_costs) as Word;
-        let gas_used_by_signature_checks = tx.gas_used_by_signature_checks(gas_costs);
-        let gas_used_by_predicates = tx.gas_used_by_predicates();
-        let gas_limit = tx.limit();
-        let gas_price = tx.price();
+        let min_gas = tx.min_gas(gas_costs, params);
+        let max_gas = tx.max_gas(gas_costs, params);
+        let min_fee = tx.min_fee(gas_costs, params).try_into().ok()?;
+        let max_fee = tx.max_fee(gas_costs, params).try_into().ok()?;
 
-        Self::checked_from_values(
-            params,
-            metered_bytes,
-            gas_used_by_signature_checks,
-            gas_used_by_metadata,
-            gas_used_by_predicates,
-            gas_limit,
-            gas_price,
-        )
+        if min_fee > max_fee {
+            return None
+        }
+
+        Some(Self::new(min_fee, max_fee, min_gas, max_gas))
     }
 }
 
+fn gas_to_fee(gas: Word, gas_price: Word, factor: Word) -> u128 {
+    let total_price = (gas as u128)
+        .checked_mul(gas_price as u128)
+        .expect("Impossible to overflow because multiplication of two `u64` <= `u128`");
+    // TODO: use native div_ceil once stabilized out from nightly
+    num_integer::div_ceil(total_price, factor as u128)
+}
+
 /// Means that the blockchain charges fee for the transaction.
-pub trait Chargeable {
+pub trait Chargeable: field::Inputs + field::Witnesses + field::Policies {
     /// Returns the gas price.
-    fn price(&self) -> Word;
+    fn price(&self) -> Word {
+        self.gas_price()
+    }
+
+    /// Returns the minimum gas required to start transaction execution.
+    fn min_gas(&self, gas_costs: &GasCosts, fee: &FeeParameters) -> Word {
+        let bytes_gas = self.metered_bytes_size() as u64 * fee.gas_per_byte;
+        // It's okay to saturate because we have the `max_gas_per_tx` rule for transaction
+        // validity. In the production, the value always will be lower than
+        // `u64::MAX`.
+        self.gas_used_by_inputs(gas_costs)
+            .saturating_add(self.gas_used_by_metadata(gas_costs))
+            .saturating_add(bytes_gas)
+    }
 
-    /// Returns the gas limit.
-    fn limit(&self) -> Word;
+    /// Returns the maximum possible gas after the end of transaction execution.
+    ///
+    /// The function guarantees that the value is not less than [Self::min_gas].
+    fn max_gas(&self, gas_costs: &GasCosts, fee: &FeeParameters) -> Word {
+        let remaining_allowed_witness = self
+            .witness_limit()
+            .saturating_sub(self.witnesses().size_dynamic() as u64)
```

Comment:
I was thinking `witness_limit()` returned the _count_ of witnesses. You are subtracting the _size_ of all the witnesses, so maybe I was mistaken. Just want to make sure you aren't subtracting size from count.

Reply:
`WitnessLimit` limits the total size of witnesses in bytes=) Because each witness can be arbitrary in size, limiting by the count will not help

<img width="900" alt="image" src="https://github.com/FuelLabs/fuel-vm/assets/18346821/a240f0df-7977-4431-8124-7838f905c290">